packages/effect/src/DateTime.ts

```diff
@@ -0,0 +1,1590 @@
+)
+
+/**
+ * Create a `DateTime` from a `Date`.
+ *
+ * If the `Date` is invalid, an `IllegalArgumentException` will be thrown.
+ *
+ * @since 3.6.0
+ * @category constructors
+ */
+export const unsafeFromDate = (date: Date): DateTime.Utc => {
+  const epochMillis = date.getTime()
+  if (Number.isNaN(epochMillis)) {
+    throw new IllegalArgumentException("Invalid date")
+  }
+  const self = Object.create(ProtoUtc)
+  self.epochMillis = epochMillis
+  return self
+}
+
+/**
+ * Create a `DateTime` from one of the following:
+ * - A `DateTime`
+ * - A `Date` instance (invalid dates will throw an `IllegalArgumentException`)
+ * - The `number` of milliseconds since the Unix epoch
+ *
+ * @since 3.6.0
+ * @category constructors
+ */
+export const fromInput = <A extends DateTime.Input>(input: A): DateTime.PreserveZone<A> => {
+  if (isDateTime(input)) {
+    return input as DateTime.PreserveZone<A>
+  } else if (input instanceof Date) {
+    return unsafeFromDate(input) as DateTime.PreserveZone<A>
+  }
+  return unsafeFromDate(new Date(input)) as DateTime.PreserveZone<A>
+}
+
+/**
+ * Safely create a `DateTime` from a `Date`, returning `None` if the `Date` is invalid.
+ *
+ * @since 3.6.0
+ * @category constructors
+ */
+export const fromDate: (date: Date) => Option.Option<DateTime.Utc> = Option.liftThrowable(unsafeFromDate)
+
+/**
+ * Convert a partial `DateTime.Parts` into a `DateTime`.
+ *
+ * If a part is missing, it will default to the smallest possible value. (months will start at 1, days at 1, hours at 0 etc.)
+ *
+ * @since 3.6.0
+ * @category constructors
+ */
+export const fromParts = (parts: Partial<Exclude<DateTime.Parts, "weekDay">>): DateTime.Utc => {
+  const date = new Date(0)
+  date.setUTCFullYear(
+    parts.year ?? 0,
+    parts.month ? parts.month - 1 : 0,
+    parts.day ?? 1
+  )
+  date.setUTCHours(
+    parts.hours ?? 0,
+    parts.minutes ?? 0,
+    parts.seconds ?? 0,
+    parts.millis ?? 0
+  )
+  return unsafeFromEpochMillis(date.getTime())
+}
+
+/**
+ * Parse a string into a `DateTime`, using `Date.parse`.
+ *
+ * @since 3.6.0
+ * @category constructors
+ */
+export const fromString = (input: string): Option.Option<DateTime.Utc> => fromDate(new Date(input))
+
+/**
+ * Parse a string into a `DateTime`, using `Date.parse`.
+ *
+ * If the string is invalid, an `IllegalArgumentException` will be thrown.
+ *
+ * @since 3.6.0
+ * @category constructors
+ */
+export const unsafeFromString = (input: string): DateTime.Utc => unsafeFromDate(new Date(input))
+
+/**
+ * Get the current time using the `Clock` service and convert it to a `DateTime`.
+ *
+ * @since 3.6.0
+ * @category constructors
+ */
+export const now: Effect.Effect<DateTime.Utc> = Effect.map(
+  Effect.clock,
+  (clock) => unsafeFromEpochMillis(clock.unsafeCurrentTimeMillis())
+)
+
+/**
+ * Get the current time using `Date.now`.
+ *
+ * @since 3.6.0
+ * @category constructors
+ */
+export const unsafeNow: LazyArg<DateTime.Utc> = () => unsafeFromEpochMillis(Date.now())
```

Comment:
Why is this unsafe? Shouldn't `Date.now` always yield a numerical output that should be convertable safely?

Reply:
Because `Date.now()` is a side effect that should be an `Effect`. And not using `Clock` means testing is harder.

Reply:
Similarly, why use unsafeCurrentTimeMillis with `now` vs something like this?

```ts
export const now: Effect.Effect<Utc> = Effect.map(
  Effect.flatMap(Effect.clock, (clock) => clock.currentTimeMillis),
  (ms) => makeUtc(ms)
)
```

Reply:
We are already in an effect context here so it is fine.